packages/cache/src/lib/CacheMiddleware.ts

```diff
@@ -0,0 +1,672 @@
+        }
+
+        return false;
+      };
+
+      const { errors } = await cache[Method.Find]({ method: Method.Find, hook: cacheHook, type, path, errors: [] });
+
+      payload.errors = [...payload.errors, ...errors];
+    } else {
+      const findPayload = await cache[Method.Find]({
+        method: Method.Find,
+        type,
+        value,
+        path: ['value', ...(path ?? [])],
+        errors: []
+      });
+
+      // the second check is if the key isn't null, which would mean it wasn't found
+      if (isPayloadWithData<CacheMiddleware.Document<StoredValue>>(findPayload) && findPayload.data[0] !== null) {
+        const { data } = findPayload;
+
+        if (await this.checkNotExpired(data[1], data[0])) {
+          payload.data = [data[0], data[1].value];
+        } else {
+          const bypassPayload = await this.provider[Method.Get]({ method: Method.Get, key: data[0], path: path ?? [], errors: [] });
+
+          if (isPayloadWithData<StoredValue>(bypassPayload)) {
+            const { data: bypassData } = bypassPayload;
+
+            payload.data = [data[0], bypassData];
+          }
+        }
+      }
+
+      payload.errors = [...payload.errors, ...findPayload.errors];
+    }
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.Keys](payload: Payloads.Keys): Promise<Payloads.Keys> {
+    const { provider: cache } = this.context;
+
+    await this.populateCache();
+
+    const keysPayload = await cache[Method.Keys]({ method: Method.Keys, errors: [] });
+
+    if (isPayloadWithData(keysPayload)) {
+      const { data } = keysPayload;
+
+      payload.data = data;
+    }
+
+    const { errors } = keysPayload;
+
+    payload.errors = [...payload.errors, ...errors];
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.Values](payload: Payloads.Values<StoredValue>): Promise<Payloads.Values<StoredValue>> {
+    const { provider: cache } = this.context;
+
+    await this.populateCache();
+
+    const valuesPayload = await cache[Method.Values]({ method: Method.Values, errors: [] });
+
+    if (isPayloadWithData<CacheMiddleware.Document<StoredValue>[]>(valuesPayload)) {
+      const { data } = valuesPayload;
+
+      payload.data = data.map(({ value }) => value);
+    }
+
+    const { errors } = valuesPayload;
+
+    payload.errors = [...payload.errors, ...errors];
+
+    return payload;
+  }
+
+  public async [Method.Map]<Value = StoredValue>(payload: Payloads.Map.ByHook<StoredValue, Value>): Promise<Payloads.Map.ByHook<StoredValue, Value>>;
+  public async [Method.Map]<Value = StoredValue>(payload: Payloads.Map.ByPath<Value>): Promise<Payloads.Map.ByPath<Value>>;
+  @PreProvider()
+  public async [Method.Map]<ReturnValue = StoredValue>(
+    payload: Payloads.Map<StoredValue, ReturnValue>
+  ): Promise<Payloads.Map<StoredValue, ReturnValue>> {
+    const { provider: cache } = this.context;
+    const { hook, path, type } = payload;
+
+    if (hook && isMapByHookPayload(payload)) {
```

Comment:
```suggestion
    if (isMapByHookPayload(payload)) {
```
Same as before regarding destructuring inside the scope.